File: testsuite/tests/ada/0015-controlled/cpp/main.cpp
#include "polyglot_ada_arrays.h"
#include "test.h"

class InheritsControlled : public test::Cont { };

int main() {
    {
        test::Cont cont;
        test::Cont cont2;
        cont = cont2;
        test::Cont foo{cont};
        cont.p();
        cont2.p();
        foo.p();
    }
    std::cout << std::endl;
    {
        InheritsControlled cont;
        cont.p();
    }
    std::cout << std::endl;
    {
        test::Cont cont;
        test::ContWrapper wrapper(cont);
        test::ContWrapper cont2;
        cont.p();
        wrapper.get_c()->p();
        cont2.get_c()->p();
    }
    std::cout << std::endl;
    {
       polyglot::ada::arrays::polyglot_array<test::Cont> arr = test::get_arr();
       test::foo(arr);
       test::foo(arr);
       test::foo(arr);
    }
    {
        test::Cont c1;
        test::Cont c2;
        std::cout << (c1 == c2) << "\n";
    }
}

File: testsuite/tests/ada/0015-controlled/java/Main.java
import com.adacore.libtest.test.TestPackage;
import com.adacore.libtest.test.Cont;
import com.adacore.libtest.test.ContWrapper;

public class Main {

    public static void main(String[] args) throws Throwable {
        try (
            Cont cont = new Cont();
            Cont cont2 = new Cont()
        ) {
            cont.copyFrom(cont2);
            try (Cont foo = new Cont(cont)) {
                cont.p();
                cont2.p();
                foo.p();
            }
        }
        System.out.println();

        try(
            Cont cont = new Cont();
            ContWrapper wrapper = new ContWrapper(cont);
            ContWrapper cont2 = new ContWrapper();
        ) {
            cont.p();
            wrapper.getC().p();
            cont2.getC().p();
        }
        System.out.println();

        try(
            Cont.Array arr = TestPackage.getArr();
        ) {
            TestPackage.foo(arr);
            TestPackage.foo(arr);
            TestPackage.foo(arr);
        }

        try(
            Cont c1 = new Cont();
            Cont c2 = new Cont()
        ) {
            System.out.println(c1.operatorEq(c2));
        }
        System.out.println();

        try(
            Cont cont = TestPackage.getCont();
            Cont cont2 = new Cont(cont);
        ) {
            cont.p();
            cont2.p();
        }
    }
}
